src/common/types/validity_mask.cpp

```diff
@@ -96,6 +96,39 @@ void ValidityMask::SliceInPlace(const ValidityMask &other, idx_t target_offset,
 		memcpy(target_validity + target_offset_entries, source_validity + source_offset_entries,
 		       sizeof(validity_t) * EntryCount(count));
 		return;
+	} else if (IsAligned(target_offset)) {
+		//	Simple common case where we are shifting into an aligned mask (e.g., 0 in Slice above)
+		const idx_t entire_units = count / BITS_PER_VALUE;
+		const idx_t ragged = count % BITS_PER_VALUE;
+		const idx_t tail = source_offset % BITS_PER_VALUE;
+		const idx_t head = BITS_PER_VALUE - tail;
+		auto source_validity = other.GetData() + (source_offset / BITS_PER_VALUE);
+		auto target_validity = this->GetData() + (target_offset / BITS_PER_VALUE);
+		auto src_entry = *source_validity++;
+		for (idx_t i = 0; i < entire_units; ++i) {
+			//	Start with head of previous src
+			validity_t tgt_entry = src_entry >> tail;
+			src_entry = *source_validity++;
+			// 	Add in tail of current src
+			tgt_entry |= (src_entry << head);
+			*target_validity++ = tgt_entry;
+		}
+		//	Finish last ragged entry
+		if (ragged) {
+			//	Start with head of previous src
+			validity_t tgt_entry = (src_entry >> tail);
+			//  Add in the tail of the next src, if head was too small
+			if (head < ragged) {
+				src_entry = *source_validity++;
```

Comment:
Is load of 64 bits of source_validity here always valid? This should boil down to whether vector data is allocated aligned AND in multiple of 64bits.

IFF this is problematic, same solution but cutting the loop short at (entire_units - 1) iterations, and finishing off with `Set( offset1....,other.RowIsValid(offset2));` should work

Reply:
The caller has requested bits in the next source entry, so we have to assume it is valid to read.

Reply:
What I meant is: if source it's like 15 bits long, it's valid to read 64 (= BITS_PER_VALUE) of them?
Or alternatively, but it's basically the same problem, if we are reading the last few validity bits, can we actually load a few bits/bytes over the boundary?

Reply:
Yes because the buffers are allocated to contain 64 bit entries.  And we mask off the bits we don't want just below.